pkg/controller/deployment/deployment_controller_test.go

```diff
@@ -286,3 +295,218 @@ func filterInformerActions(actions []core.Action) []core.Action {
+	f := newFixture(t)
+	now := metav1.Now()
+	later := metav1.Time{Time: now.Add(time.Minute)}
+
+	foo := newDeployment("foo", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	foo.CreationTimestamp = now
+	bar := newDeployment("bar", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	bar.CreationTimestamp = later
+
+	f.dLister = append(f.dLister, foo, bar)
+	f.objects = append(f.objects, foo, bar)
+
+	f.expectUpdateDeploymentStatusAction(bar)
+	f.run(getKey(bar, t))
+
+	actions := f.client.Actions()
+	d := actions[0].(core.UpdateAction).GetObject().(*extensions.Deployment)
+	if len(d.Annotations[util.OverlapAnnotation]) == 0 {
+		t.Errorf("annotations werent updated for the overlapping deployment: %v", d.Annotations)
+	}
+}
+
+// TestSyncNonOverlappingDeployment ensures that from two overlapping deployments, the older
+// one will be synced and the newer will be marked as overlapping. Note that in reality it's
+// not always the older deployment that is the one that works vs the rest but the one which
+// has the selector unchanged for longer time.
+func TestSyncNonOverlappingDeployment(t *testing.T) {
+	f := newFixture(t)
+	now := metav1.Now()
+	later := metav1.Time{Time: now.Add(time.Minute)}
+
+	foo := newDeployment("foo", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	foo.CreationTimestamp = now
+	bar := newDeployment("bar", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	bar.CreationTimestamp = later
+
+	f.dLister = append(f.dLister, foo, bar)
+	f.objects = append(f.objects, foo, bar)
+
+	f.expectUpdateDeploymentStatusAction(bar)
+	f.expectCreateRSAction(newReplicaSet(foo, "foo-rs", 1))
+	f.expectUpdateDeploymentAction(foo)
+	f.expectUpdateDeploymentStatusAction(foo)
+	f.run(getKey(foo, t))
+
+	actions := f.client.Actions()
+	d := actions[0].(core.UpdateAction).GetObject().(*extensions.Deployment)
+	if len(d.Annotations[util.OverlapAnnotation]) == 0 {
+		t.Errorf("annotations werent updated for the overlapping deployment: %v", d.Annotations)
+	}
+}
+
+// TestDeletedDeploymentShouldCleanupOverlaps ensures that the deletion of a deployment
+// will cleanup any deployments that overlap with it.
+func TestDeletedDeploymentShouldCleanupOverlaps(t *testing.T) {
+	f := newFixture(t)
+	now := metav1.Now()
+	earlier := metav1.Time{Time: now.Add(-time.Minute)}
+	later := metav1.Time{Time: now.Add(time.Minute)}
+
+	foo := newDeployment("foo", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	foo.CreationTimestamp = earlier
+	foo.DeletionTimestamp = &now
+	bar := newDeployment("bar", 1, nil, nil, nil, map[string]string{"foo": "bar"})
+	bar.CreationTimestamp = later
+	bar.Annotations = map[string]string{util.OverlapAnnotation: "foo"}
+
+	f.dLister = append(f.dLister, foo, bar)
+	f.objects = append(f.objects, foo, bar)
+
+	f.expectUpdateDeploymentStatusAction(bar)
+	f.expectUpdateDeploymentStatusAction(foo)
+	f.run(getKey(foo, t))
+
+	for _, a := range f.client.Actions() {
+		action, ok := a.(core.UpdateAction)
```

Comment:
I am not gaining really anything by doing this. I still have a slice of actions that I need to traverse, cast to UpdateAction, get the object, and check the annotation.